File: src/lowlevel/server/icefsoperators/icefsGetXattr.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 03:50:05
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsGetXattr.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"fmt"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsGetXattr(fakeInode uint64, name string, size uint64) (status int32, valueSize int64, value []byte) {
	var procName string
	var err error
	var sz int
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	procName = fmt.Sprintf("/proc/self/fd/%v", inode.fd)
	inode.inodeLock.RUnlock()
	value = make([]byte, size)
	sz, err = syscall.Getxattr(procName, name, value)
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}
	status = icefserror.ICEFS_EOK
	valueSize = int64(sz)
	return
}

func (s *IcefsGRpcServer) DoIcefsGetXattr(ctx context.Context, req *pb.IcefsGetXattrReq) (*pb.IcefsGetXattrRes, error) {
	var res pb.IcefsGetXattrRes

	res.Status, res.Size, res.Value = s.server.doIcefsGetXattr(req.Inode, req.Name, req.Size)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsGetXattr(ctx context.Context, req *icefsthrift.IcefsGetXattrReq) (*icefsthrift.IcefsGetXattrRes, error) {
	var res icefsthrift.IcefsGetXattrRes

	res.Status, res.Size, res.Value = s.server.doIcefsGetXattr(uint64(req.Inode), req.Name, uint64(req.Size))

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsGetXattr.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:48
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsGetXattr.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsGetXattr(fuse_req_t fuseReq, fuse_ino_t inode,
                                  const char *name, size_t size) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsGetXattrReq req;
      icefsgrpc::IcefsGetXattrRes res;
      grpc::ClientContext ctx;

      req.set_inode(inode);
      req.set_name(name);
      req.set_size(size);

      grpc::Status status = gRpcClient->DoIcefsGetXattr(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        if (size) {
          int32_t ret =
              fuse_reply_buf(fuseReq, res.value().c_str(), res.size());
          if (ret != ICEFS_EOK) {
            fuse_reply_err(fuseReq, ret);
            std::cout << "ERR:" << __FUNCTION__ << "fuse_reply_buf" << ret
                      << std::endl;
          }
        } else {
          fuse_reply_xattr(fuseReq, res.size());
        }
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsGetXattrReq req;
      icefsthrift::IcefsGetXattrRes res;

      req.__set_inode(inode);
      req.__set_name(name);
      req.__set_size(size);
      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsGetXattr(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        if (size) {
          int32_t ret = fuse_reply_buf(fuseReq, res.value.c_str(), res.size);
          if (ret != ICEFS_EOK) {
            fuse_reply_err(fuseReq, ret);
            std::cout << "ERR:" << __FUNCTION__ << "fuse_reply_buf" << ret
                      << std::endl;
          }
        } else {
          fuse_reply_xattr(fuseReq, res.size);
        }
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}